types/reactstrap/lib/Alert.d.ts

```diff
@@ -8,6 +8,7 @@ export interface UncontrolledProps extends React.HTMLAttributes<HTMLElement> {
   transitionAppearTimeout?: number;
   transitionEnterTimeout?: number;
   transitionLeaveTimeout?: number;
+  [others: string]: any;
```

Comment:
This is needed for all types that have the tag prop as tag can take any React element, meaning any additional prop

Reply:
this kind of makes types useless, no?  I don't think this is a good thing to add this to the definitions.  

What attribute is missing?  this already extends `React.HTMLAttributes<HTMLElement>`

Reply:
This only works if you use standard html elements as tags but not for custom react components. Say for example I want to use `<my-component my-attr />` as tag then passing my-attr will be marked as error bt Typescript. Since tag can be React components as well this becomes mandatory. Its. If you have another way to tackle this I'm happy to hear about it. Originally I was facing the issue in the button components as I wanted them to use Link from react router as tag. But then it made sense that all tags are same

Reply:
One possible solution is to use mapped types - this would require all props to be types instead of interfaces though....

Here's an example with ButtonProps:

```
export type ButtonProps<TagProps ={}> = React.HTMLProps<HTMLButtonElement> & {
  outline?: boolean;
  active?: boolean;
  block?: boolean;
  color?: string;
  disabled?: boolean;
  tag?: React.ReactType<TagProps>;
  innerRef?: string | ((instance: HTMLButtonElement) => any);

  onClick?: React.MouseEventHandler<any>;
  size?: any;
  id?: string;
  style?: React.CSSProperties;

  cssModule?: CSSModule;
} & { 
  [P in keyof TagProps]: TagProps[P];
}
```

Then to use typed ButtonProps, you would have to declare your own Button:

```
declare const MyButton: React.StatelessComponent<ButtonProps<LinkProps>>
```

Reply:
IMO it would make using the types more complex as I would have to create and maintain my own types for all the combinations I use with a tag. The only downside I see to my approach is that you wouldn't get a warning if you add a non-existing type but since the element supports tag you can add anything as an attribute that is supported by the tag which is the expected behavior of the component and the Proptypes defined for those

Reply:
For the majority of developers who are not using custom tags, I believe this is the best solution as it doesn't add any overhead/complexity.

For the users who are using custom tags, they will just have to declare the tag props they want to use.  It doesn't make it any more complex than any other types that you have to manage.  

The whole point of typescript is to have strong types, and most of the times, that means declaring your own generic types for your custom components.  This is similar to having to maintain your React components with strongly typed Props. Sure, you could declare your components without strongly typing your props:
```class MyComponent extends React.Component<any> { }```
but why use typescript at all then?

The downside to your approach is that it allows anything to be a prop and breaks the strong type-ness of the components.

Reply:
As much as I dislike having a `[key: string]: any` anywhere, I think it really is the correct call here. Even if most developers aren't using custom tags, there's no reason to make it more difficult for those that are. I don't see how this breaks strong types for the props that reactstrap actually grabs, and this is hardly the first place `[key: string]: any` is used in type definitions.

Tim, your suggestion also doesn't address if the component rendering the reactstrap component is also getting the tag / extra props dynamically.

My only question about this is a semantic question. It seems most people choose to use to use `[key: string]: any` over `[others: string]: any` - is there a specific reason you used `others` instead of `key`?

There's also an open issue on this: #20461.

Reply:
I agree with @FaithForHumans 
IMO, static typing should be helping you and not just bloat your code, as everyone is not Type nazis, I wouldn't want to create complex custom types for the React strap components in my codebase to use tags, If doing that is the only way to use this type definitions than I would rather go without it.

@FaithForHumans I also normally prefer `[key: string]: any` but I saw that the definitions already had few interfaces with `[others: string]: any` so I just followed the same to be consistent.

Reply:
@timc13 To give you a real example, in my codebase I use Reactstrap extensively and in most cases with custom componenets as tags, so going with your approach woukd mean I would have to create numerous custom types for all the combinations I have which is a real overhead for me and my team, and something I'm against as it makes the benefits of static typing less appealing compared to the overhead. I love JS as it is and static typing is just a convenience as long as it doesnt add bloat.
Also we use Reactstrap in [JHipster](https://www.jhipster.tech/) extensively and I wouldnt like to bloat it with unnecessary custom types

Reply:
Declaring a generic component is not the same as "creating complex custom types"

Sure, there are plenty of libraries that use `any`, but I've found those to be poorly maintained and less useful than well typed libraries that avoided it.  By introducing it here, we risk poor maintenance updates as any new prop just falls into `any`

On the flip side, there are plenty of libraries that do maintain strict typing - here is an example where you have to declare a generic component before using it: https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react-autosuggest

Reply:
[Your example uses the same `[key: string] any` notation for the react props](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/27e6a748e97e4b485d550cd515c31647ea792cf0/types/react-autosuggest/index.d.ts#L60). Reactstrap is structured differently, that's why this suggestion occurs in the base interface.

----------

I see the pros/cons of doing `[key: string] any` and having to create generic components. That being said, I'm, going to try and offer a solution for you @deepu105 if we did require declaring a custom generic component. If we did something like:

    export type RowProps<T> =  React.HTMLProps<HTMLElement> & T & {
        ...
    }
    declare class Row<T extends object = {}> extends React.Component<RowProps<T>> {}
    export default Row;

would you find it reasonable for you to make a reactstrap generics utility file that you can import where fit and defines all of the components you need like:

    const GenericRow = reactstrap.Row as { new (): reactstrap.Row< {[key: string]: any} > }; 

That doesn't require you to create all of your types if you don't want to, and you still get the same functionality with what you've offered in this PR.

I would suggest including all of the "generics" here, but I don't know of a way to include them without them being shadowed.

------------

Sidebar: As to the `[others: string]` vs `[key: string]`, it looks like [~2](https://github.com/DefinitelyTyped/DefinitelyTyped/search?utf8=%E2%9C%93&q=%22%5Bkey%3A+string%5D%3A+any%22&type=) (both from the current reactstrap typings, which I didn't notice) versus [~273](https://github.com/DefinitelyTyped/DefinitelyTyped/search?utf8=%E2%9C%93&q=%22%5Bkey%3A+string%5D%3A+any%22&type=) results. Personally, I don't care about that part either way, I was just curious.

Reply:
@FaithForHumans Thanks for the suggestion. IMO thats still unnecessary bloat. I personally dont like to do these kind of things for the sake of a lib as it doesnt add much value to my codebase. Personally the only reason for me to use typings for a lib is for code assistance in editor/IDE for the componenets, which I would get even with the `[key: string] any` approach. Doing all these generic business for the sake of doing strong typing is an overkill IMO and these kind of things are the ones that scares away people from using types in JS. having said that I also undesratnd and agree that hvaing strong types reduces errors like typos in attribute names and having unnecessary/unsupported attributes but i'm personally ok with. 
So for the sake of others who might use this lib, I guess I can live with such decalrations in a util file as long as I dont have to do so much duct taping so that I could use this just for getting some auto suggestions for prop names. So as a fellow community maintainer I suggest you do what is good for the community rather than what is good for me. I can live without using the libraray as well.